src/models/event.js

```diff
@@ -776,9 +776,23 @@ utils.extend(module.exports.MatrixEvent.prototype, {
      * @param {Object} event the object to assign to the `event` property
      */
     handleRemoteEcho: function(event) {
+        const oldUnsigned = this.getUnsigned();
+        const oldId = this.getId();
         this.event = event;
+        // keep being redacted if
+        // the event was redacted as a local echo
+        if (oldUnsigned.redacted_because) {
+            if (!this.event.unsigned) {
```

Comment:
So local echo for redaction can't use `makeRedacted` because that clears all the event properties as you mention, and local echo for redaction needs to be undo-able in case the event gets cancelled. So instead it just sets redacted_because and a flag to prevent it from returning the content. `makeRedacted` is only used once we are sure the event can't become "unredacted" anymore, e.g. when the redaction comes in from /sync.

https://github.com/vector-im/riot-web/issues/10006 does seem to indicate that we're not always doing the right thing wrt to redactions, and re-exposing props, but need to think more about what's actually going wrong there.

Does that clear it up?

Reply:
Hmm... if the local echo was redacted, we can't use `makeRedacted`, I agree. But in this code here, we are receiving the remote echo, so there's no chance of cancelling, I am pretty sure...? So, could we use `makeRedacted` for this bit, or is there still a different case...?

Oh! I guess the issue is not about the _original_ event being cancelled, but instead the redaction of the local echo might be cancelled? 😭 Okay, if that's it, then it seems fine enough, but maybe a longer comment explaining this will help for the future.

Reply:
Indeed, the code above handles keeping the locally redacted state while receiving the remote echo for the *redacted* event, not the *redaction* (which might still get cancelled for all we know). In other words: we already know the event is redacted (from the local echo), but sync hasn't told us so yet.

Will expand the comment saying so much :+1: